include/swift/Reflection/TypeRef.h

```diff
@@ -36,13 +37,22 @@ enum class TypeRefKind {
 #undef TYPEREF
 };
 
-#define FIND_OR_CREATE_TYPEREF(Allocator, TypeRefTy, ...) \
-  auto ID = Profile(__VA_ARGS__); \
-  const auto Entry = Allocator.template TypeRefTy##s.find(ID); \
-  if (Entry != Allocator.template TypeRefTy##s.end()) \
-    return Entry->second; \
-  const auto TR = Allocator.template makeTypeRef<TypeRefTy>(__VA_ARGS__); \
-  Allocator.template TypeRefTy##s.insert({ID, TR}); \
+// MSVC reports an error if we use "template"
+// Clang reports an error if we don't use "template"
+#if defined(__clang__) || defined(__GNUC__)
+#define DEPENDENT_TEMPLATE template
+#else
+#define DEPENDENT_TEMPLATE
+#endif
+
+#define FIND_OR_CREATE_TYPEREF(Allocator, TypeRefTy, ...)                      \
+  auto ID = Profile(__VA_ARGS__);                                              \
+  const auto Entry = Allocator.DEPENDENT_TEMPLATE TypeRefTy##s.find(ID);       \
+  if (Entry != Allocator.DEPENDENT_TEMPLATE TypeRefTy##s.end())                \
+    return Entry->second;                                                      \
+  const auto TR =                                                              \
+      Allocator.DEPENDENT_TEMPLATE makeTypeRef<TypeRefTy>(__VA_ARGS__);        \
+  Allocator.DEPENDENT_TEMPLATE TypeRefTy##s.insert({ID, TR});                  \
```

Comment:
We should file a SR to address this difference and try to unify the behavior once the compilers are updated.